src/psyclone/psyir/tools/definition_use_chains.py

```diff
@@ -0,0 +1,629 @@
+            # TODO Work out if symbol is local to the scope or
+            # globally accessible. Anything that isn't a local
+            # variable in a Routine has to be assumed as globally
+            # accessible for now.
+            self._is_local = isinstance(
+                reference.symbol.interface, AutomaticInterface
+            )
+        else:
+            self._is_local = is_local
+
+        # The uses, defsout and killed sets as defined for each basic block.
+        self._uses = []
+        self._defsout = []
+        self._killed = []
+
+        # The output map, mapping between nodes and the reach of that node.
+        self._reaches = []
+        self._writes = []
+
+    @property
+    def uses(self):
+        """
+        :returns: the list of nodes used by this DefinitionUseChain.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._uses
+
+    @property
+    def defsout(self):
+        """
+        :returns: the list of output nodes computed by this DefinitionUseChain.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._defsout
+
+    @property
+    def killed(self):
+        """
+        :returns: the list of killed output nodes computed by this
+                  DefinitionUseChain.
+        :rtype: List[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        return self._killed
+
+    @property
+    def is_basic_block(self):
+        """
+        :returns: whether the scope of this DefinitionUseChain is a basic
+                  block, i.e. whether it contains any control flow nodes.
+        :rtype: bool
+        """
+        # This is a basic block if there is no control flow inside this scope
+        # then it is a basic block.
+        # In PSyclone, possible control flow nodes are IfBlock, Loop
+        # and WhileLoop.
+        for node in self._scope:
+            c_f_nodes = node.walk((IfBlock, Loop, WhileLoop))
+            if len(c_f_nodes) > 0:
+                return False
+        return True
+
+    def find_forward_accesses(self):
+        """
+        Find all the forward accesses for the reference defined in this
+        DefinitionUseChain.
+        Forward accesses are all of the References (or Calls) that read
+        or write to the symbol of the reference up to the point that a
+        write to the symbol is guaranteed to occur.
+        PSyclone only assumes a write to the symbol is guaranteed to occur
+        if it occurs outside of a control flow region.
+
+        :returns: the forward accesses of the reference given to this
+                  DefinitionUseChain
+        :rtype: list[:py:class:`psyclone.psyir.nodes.Node`]
+        """
+        # FIXME If all defsout is in control flow we should add a None into
+        # the defsout array. @Reviewer not sure about this - should we make
+        # it clear somehow its not guaranteed to be written to or does it not
+        # matter?
```

Comment:
I am not sure about this, maybe we need to discuss this again with some examples/use_cases of how this will be used?

Reply:
We discussed this previously when I was on-site - and initially thought this might be a good idea, but I'm not actually sure.
It could be left up to the code that uses the result to check this, as I think some of use-cases will not care if the write is "guaranteed" or not.

Reply:
I think also until #2760 is implemented it doesn't make sense to fix this anyway.

Reply:
We have the rule that FIXMEs need to be solved during the PR or converted to TODOs with a reference to an issue, but if I grep for FIXME I still see 3 of them: there is one about an optimisation that I would probably just remove for now, but for the other two could you open a issue? and it will be very helpful to have some example code to show where the issue happens.

Unrelated to this PR but could you also s/FIXME/TODO/ in src/psyclone/dynamo0p3.py

Reply:
Actually i think the documentation already covers it so I'll remove it.